parser/agent.go

```diff
@@ -107,6 +109,108 @@ func InstrumentMain(manager *InstrumentationManager, c *dstutil.Cursor) {
 	}
 }
 
+func checkForExistingApplicationInFunctions(manager *InstrumentationManager, c *dstutil.Cursor) {
+	dstutil.Apply(c.Node(), func(cursor *dstutil.Cursor) bool {
+		checkFuncDeclForApplication(manager, cursor.Node())
+		handleAssignStmtForAgentVariable(manager, cursor.Node())
+		return true
+	}, nil)
+}
+
+// Checks return values of a given function. If the function returns a new relic application, it is marked as a "setup" function
+func checkFuncDeclForApplication(manager *InstrumentationManager, node dst.Node) bool {
+	decl, ok := node.(*dst.FuncDecl)
+	if !ok || decl.Type == nil || decl.Type.Results == nil {
+		return false
+	}
+
+	for _, result := range decl.Type.Results.List {
+		// Checking if return type of function is a new relic application
+		starExpr, ok := result.Type.(*dst.StarExpr)
+		if !ok {
+			continue
+		}
+
+		ident, ok := starExpr.X.(*dst.Ident)
+		if !ok {
+			continue
+		}
+
+		if ident.Path == codegen.NewRelicAgentImportPath && ident.Name == "Application" {
+			manager.setupFunc = decl
+			return true
+		}
+	}
+	return false
+}
+
+func handleAssignStmtForAgentVariable(manager *InstrumentationManager, node dst.Node) bool {
+	assign, ok := node.(*dst.AssignStmt)
+	if !ok {
+		return false
+	}
+
+	for pos, rhs := range assign.Rhs {
+		call, ok := rhs.(*dst.CallExpr)
+		if !ok {
+			continue
+		}
+
+		ident, ok := call.Fun.(*dst.Ident)
+		if !ok || ident.Obj == nil {
+			continue
+		}
+
+		funcCall, ok := ident.Obj.Decl.(*dst.FuncDecl)
+		if !ok || manager.setupFunc != funcCall {
+			continue
+		}
+
+		// This is our setup function. We can now get the appName!
+		if ident, ok := assign.Lhs[pos].(*dst.Ident); ok {
+			manager.agentVariableName = ident.Name
+			return true
+		}
+	}
+	return false
+}
+
+// checkForExistingApplicationInMain checks for existing application in main.
+// If an application is detected in a function inside of main, we mark that one
```

Comment:
> If an application is detected in a function inside of main

Does this mean a function literal _inside_ of main, or just anywhere inside of the `main` function?